Adds custom reporter and logic to generate conformance docs and list of tests

johnSchnake opened the pull request:
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**:
The existing walk.go and conformance.txt have a few shortcomings
which we'd like to resolve:
 - difficult to get the full test name due to test context nesting
 - complicated AST logic and understanding necessary due to the
different ways a test can be invoked and written

This changes the AST parsing logic to be much more simple and simply
looks for the comments at/around a specific line. This file/line
information (and the full test name) is gathered by a custom ginkgo
reporter which dumps the SpecSummary data to a file.

Also, the SpecSummary dump can, itself, be potentially useful for
other post-processing and debugging tasks.

**Which issue(s) this PR fixes**:
Fixes #84330

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
NONE
```

johnSchnake committed:
Update the conformance list and doc generation logic

The existing walk.go and conformance.txt have a few shortcomings
which we'd like to resolve:
 - difficult to get the full test name due to test context nesting
 - complicated AST logic and understanding necessary due to the
different ways a test can be invoked and written

This changes the AST parsing logic to be much more simple and simply
looks for the comments at/around a specific line. This file/line
information (and the full test name) is gathered by a custom ginkgo
reporter which dumps the SpecSummary data to a file.

Also, the SpecSummary dump can, itself, be potentially useful for
other post-processing and debugging tasks.

Signed-off-by: John Schnake <jschnake@vmware.com>

johnSchnake commented on test/conformance/walk.go:
```diff
@@ -14,415 +14,294 @@ See the License for the specific language governing permissions and
+	// filenames have `/go/src/k8s.io` prefix which dont exist locally
+	for i, v := range frames {
+		frames[i].File = strings.Replace(v.File,
+			"/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes",
```
I need some additional 👀 on this part; this was hacked together based on my own work but I'm not sure if I can always expect the same prefix or what I should be using to strip this prefix as appropriate.

spiffxp replied:
I will try to get this running locally and report back

johnbelamaric replied:
@spiffxp did you try this? It looks fishy.

johnSchnake commented:
# Flow

## Check out my branch

    cd ~/go/src/k8s.io/kubernetes
    
    GITHUBUSER=johnSchnake
    BRANCH=documenterReporter
    
    git remote add $GITHUBUSER http://github.com/$GITHUBUSER/kubernetes.git
    git fetch -a $GITHUBUSER
    git checkout -b $BRANCH $GITHUBUSER/$BRANCH

## FYI using this bazel version

    cd ~/go/src/k8s.io/kubernetes
    curl -L --output bazel https://github.com/bazelbuild/bazel/releases/download/0.29.1/bazel-0.29.1-linux-x86_64
    chmod +x bazel

## Bazel target (updated): list-conformance-tests

    cd ~/go/src/k8s.io/kubernetes
    ./bazel build //test/conformance:list_conformance_tests

## Confirm output

    cat  bazel-bin/test/conformance/conformance.txt|yq -r .[0:2]

    [
      {
        "testname": "Pod Lifecycle, post start exec hook",
        "codename": "[k8s.io] Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart exec hook properly [NodeConformance] [Conformance]",
        "description": "When a post start handler is specified in the container lifecycle using a ‘Exec’ action, then the handler MUST be invoked after the start of the container. A server pod is created that will serve http requests, create a second pod with a container lifecycle specifying a post start that invokes the server pod using ExecAction to validate that the post start is executed.",
        "release": "v1.9",
        "file": "test/e2e/common/lifecycle_hook.go"
      },
      {
        "testname": "Pod Lifecycle, post start http hook",
        "codename": "[k8s.io] Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart http hook properly [NodeConformance] [Conformance]",
        "description": "When a post start handler is specified in the container lifecycle using a HttpGet action, then the handler MUST be invoked after the start of the container. A server pod is created that will serve http requests, create a second pod with a container lifecycle specifying a post start that invokes the server pod to validate that the post start is executed.",
        "release": "v1.9",
        "file": "test/e2e/common/lifecycle_hook.go"
      }
    ]

## Bazel target (new): gen-conformance-docs

     cd ~/go/src/k8s.io/kubernetes
     ./bazel build //test/conformance:gen_conformance_docs
    bazel-bin/test/conformance/conformance.md

[tests/conformance/conformance.md](file://~/go/src/k8s.io/kubernetes/bazel-bin/test/conformance/conformance.md)

johnSchnake commented:
/test pull-kubernetes-e2e-gce

johnSchnake commented:
The pull-kubernetes-conformance-image-test is currently broken in general; nothing about this PR: https://prow.k8s.io/job-history/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-conformance-image-test

johnSchnake commented:
/assign @hh

fejta commented:
/assign cheftako spiffxp
/uncc
/retest

johnSchnake commented:
cc @johnbelamaric

hh commented:
/assign @johnbelamaric

hh commented:
/retest

spiffxp commented on test/conformance/walk.go:
```diff
@@ -14,415 +14,294 @@ See the License for the specific language governing permissions and
-	totalConfTests, totalLegacyTests, missingComments int
-
-	// If a test name contains any of these tags, it is ineligble for promotion to conformance
-	regexIneligibleTags = regexp.MustCompile(`\[(Alpha|Feature:[^\]]+|Flaky)\]`)
```
I have concerns this means we're no longer preventing invalid tests from being promoted to conformance

johnSchnake replied:
I don't recall removing this part but wouldn't be opposed since the promotion to conformance is a manual process that requires that the same people who administer this list sign off on the promotion.

The only time I've seen a test desire to be promoted that hit this regexp, the regexp was changed to allow the test, not the other way around. It also would motivate situations where you just remove the offending tag in order to promote to work around the issue. 

It also doesn't catch situations where the test lacks the tag in question, even though it would apply (e.g. are all tests properly marked as alpha/feature/flaky). Again, my largest experience with this was that tests that were disruptive weren't marked as such, so the regexp (which included disruptive at the time) didn't block it though it should have.

If you feel strongly I think this would be a pretty simple thing to add back in.

johnSchnake replied:
Talked about this is the conformance-wg; I will add back in this behavior.

spiffxp commented on test/conformance/BUILD:
```diff
@@ -31,22 +38,35 @@ filegroup(
-    outs = ["conformance.txt"],
-    cmd = "./$(location :conformance) $(locations //test/e2e:all-srcs) > $@",
+    outs = ["specsummaries.json"],
+    cmd = "$(location //vendor/github.com/onsi/ginkgo/ginkgo) --dryRun=true --focus=[Conformance] $(location //test/e2e:e2e.test_binary) -- --spec-dump $$(pwd)/$@ > /dev/null",
```
I think we could simulate excluding invalid tags by adding a skip regex here

johnSchnake replied:
If you did that, it would not dump their test info leading to them not resulting in the final doc. However, it wouldn't stop the test from actually being executed with `ConformanceIt(...` which would mean it could lead to the tests chosen by `--focus=[Conformance]` not lining up with the tests described in the doc generated by this.

spiffxp commented on test/conformance/walk.go:
```diff
@@ -14,415 +14,294 @@ See the License for the specific language governing permissions and
-		FileSet: token.NewFileSet(),
-	}
+func isConformance(spec *types.SpecSummary) bool {
+	return strings.Contains(getTestName(spec), "[Conformance]")
```
we could also exclude invalid tags here

spiffxp reviewed (commented):
I'll try testing this out locally later. I had taken a crack at simplifying the AST nonsense, so as to try folding in linting, but if this already there and simpler I'm fine with this approach.

My concerns are:
- how this will be used to gate tests with invalid tags
- (eventually) how this will be used to gate tests that don't already have their behaviors described

johnSchnake commented on test/conformance/walk.go:
```diff
@@ -14,415 +14,294 @@ See the License for the specific language governing permissions and
-
-		description := v.getDescription(at.Value)
-		err := validateTestName(description)
+// TODO(SCHNAKE) 1-2 tests for each case would be good
```
Remove todo; have to double check if I made the intended tests or not (its been a little while since I looked at this; I think I was done though)

johnSchnake commented on test/conformance/walk.go:
```diff
@@ -14,415 +14,294 @@ See the License for the specific language governing permissions and
-	matches := regexIneligibleTags.FindAllString(s, -1)
-	if matches != nil {
-		return fmt.Errorf("'%s' cannot have invalid tags %v", s, strings.Join(matches, ","))
+// TODO(SCHNAKE) TEST if I split this up to first get the comment map then pass comment map/stack trace to pick out the right one that would be valuable
```
Remove todo; not sure if more test work is required. Will double check.

spiffxp commented:
/retest

spiffxp commented:
the bazel test job is failing because a new conformance test was added since this PR was opened, and it's not in the yaml file, so I think this answers part of my question

spiffxp commented:
As written this will require human review that the `[Conformance]` tag was not manually added, and that `ConformanceIt` was used instead.  I manually added the tag to an `It` case, and it was picked up for inclusion in `conformance.yaml`. Do we still care about requiring the use of `ConformanceIt`?

johnSchnake commented:
> Do we still care about requiring the use of `ConformanceIt`?

I wasn't around when that went into effect; my guess is that it was added just to support the ast walker?

I don't have any reason why we would continue to care.  Literally the only thing it does is add the conformance tag so I suppose the only thing it _could_ do is prevent a typo but then it wouldn't be picked up. Likewise, every other tag would suffer that same issue so it doesn't seem to me like we need the requirement.

spiffxp commented:
I'm fine getting rid of / not requiring `ConformanceIt` if others are fine with it and we move forward with this approach. It was definitely for the benefit of AST walking.

Having walked through a few examples locally I think this does what it says on the tin.

However, I do feel like this approach is slower and more fiddly.  Now we're combining stack trace walking and AST walking.  AST walking is mainly for the benefit of picking up comments. Where could these live such that we wouldn't need AST walking in the future?

johnSchnake commented:
> Where could these live such that we wouldn't need AST walking in the future?

Great question; I can't find any way for the test binary to include info about comments. As a result it would have to be in actual code. That would mean that we'd need to do something like our own custom dry run where we could surface test data if desired.

We could move to something like passing those test details (framework, test name, description, etc) as parameters. Then if some TestContext flag was set, we would just spit out data and exit the test instead of running it.

This would also give us the means to enforce things rules like "must have non-empty/unique test names and descriptions" in the framework itself.

Could dual support the approaches for a while as we make the changes then have one PR remove the unnecessary comments.

Thoughts?

johnbelamaric commented:
I am thinking this may conflict with what we are planning around the behavior stuff. I think processing the tests to figure out the behaviors is going about it in the wrong direction. I just pushed a WIP #85960 that starts to lay out the behavior definitions. PTAL at that.

This code could be used to generate behavior YAMLs from existing tests, though I am not entirely sure it's something we want to do.

johnSchnake commented:
@johnbelamaric so that WIP shows that we'd aggregate the behavior IDs and descriptions etc in an external YAML file. But at some point we still have to tie that to code and so we face the same problem dont we? 

We can use comments to mark a test as satisfying a behavior.  If we do that the comments are available in any binary so we need to use an AST walker to connect the two. (Or are we going to do something as simple as just grepping the codebase to see that _somewhere_ that behaviorID gets used and assume it connects to a reasonable test to avoid using AST?)

If we put it in the code we can perform our own single-pass processing logic (as opposed to this PR which uses the ginkgo dry-run capabilities + a bit of AST to get the comments)

I could imagine something like:

```
framework.ConformanceIt("should do something", <behaviorID>, func() {
...
```

Then in ConformanceIt we can have logic so that we either run the test as normal or, if generating docs and such, spit out information about the test, check the YAML file to see the behaviorID is defined and is correct based on the name, etc etc.

johnbelamaric commented:
Yeah, that could work. My earliest prototype way back did something like that. Though I think a single test would often cover multiple behaviors so we need to consider the params in that context (for example, create a pod, delete a pod is two behaviors but for efficiency maybe in one test). But in code is good. We definitely want to map back to the specific test that covers a behavior, because during a run we need to say whether a given behavior works properly.

If a test covers multiple behaviors, using the comments or ConformanceIt as described above, then failing that test would mark all those behaviors as failing, even if it was only some of them. I am not sure if we care. Right now we might not, but eventually the behaviors would be classified into profiles (base, windows, etc.), and then maybe it we care?

If we want to solve that problem, then we need to have a separate test case or we need to somehow be able to use a "Describe" or something to say whether each behavior worked. Is that worth the effort?

johnbelamaric commented:
@johnSchnake just talking here with @jefftree and we think we should move forward with this and we can tweak it a bit from there

johnSchnake commented:
- Added back the regexp check for invalid tags
- rebased
- removed some debug comments of mine

Also manually confirmed the yaml file produced has the same number of conformance tests (280) as the conformance.txt

johnSchnake commented:
/retest

johnSchnake commented:
So some of the conformance test docs/names/etc had been updated since I rebased last which is what was causing the repeated test failure. It took me longer than it should have to realize that; I thought I had accidentally introduced changes instead of _needing_ to update them.  Just rebased and presume it should go green now.

johnSchnake commented:
/retest

johnSchnake commented:
/retest

johnSchnake commented:
3 changes to conformance.txt in the past day have caused rebasing slow-downs. That coupled with gce flakes has kept this PR from being ready to merge. Hoping things go green and approvals follow soon. No known issues still.

spiffxp commented on test/conformance/cf_header.md:
```diff
@@ -15,26 +15,23 @@ clean up the tests.
-Note this test needs to be fixed to also test for stderr
+- Added to conformance in release v1.13
+- Defined in code as: [k8s.io] Kubelet when scheduling a busybox command in a pod should print the output to logs [NodeConformance] [Conformance]
+By default the stdout and stderr from the process being executed in a pod MUST be sent to the pod's logs.
```
Formatting: this probably needs its own bullet, or a hard break to separate from the bullets above (currently runs on from the second bullet)

johnSchnake replied:
👍 Thanks; I'll add the hard break; that is present in the code when printing each test but not here.

spiffxp commented on test/conformance/walk.go:
```diff
@@ -14,415 +14,306 @@ See the License for the specific language governing permissions and
+			fmt.Printf("## [%s](%s)\n\n", data.TestName, data.URL)
+			fmt.Printf("- Added to conformance in release %s\n", data.Release)
+			fmt.Printf("- Defined in code as: %s\n\n", data.CodeName)
+			fmt.Printf("%s\n\n", data.Description)
```
This should have its own bullet, eg
```suggestion
			fmt.Printf("- Test description: %s\n\n", data.Description)
```

johnSchnake replied:
Do you think it looks better that way? I deliberately didnt put a bullet because in both the old and new version of the docs the description was just a text block (not a bullet). I had added the other fields as a bulleted list to differentiate them from the description.

spiffxp replied:
I'm fine with the hard break as well, my engineer brain was going for consistency with the suggestion

johnSchnake commented:
/test pull-kubernetes-e2e-gce

spiffxp commented:
/retest

johnSchnake commented:
I've seen this pull-kubernetes-verify fail on at least one other PR in this way; very strange that it gets into this loop of seeming to fail because of standard lib stuff. Can 🤞 that it just passes on the retest, but I'm not certain that there isnt a more deep issue with something in the test-infra.

spiffxp commented:
/lgtm
/approve

johnbelamaric commented:
/approve

spiffxp commented:
/retest
flake https://github.com/kubernetes/kubernetes/issues/87441

spiffxp commented:
/retest
flake #87441

johnSchnake commented:
Rebased since there was some small change in some test wording it seems.

Jefftree commented:
/lgtm
/retest

johnSchnake commented:
/retest

johnbelamaric commented:
This is taking way too long to merge. Can't be just regular flakes. Latest one shows:

```
W0205 20:25:49.216] Server terminated abruptly (error code: 14, error message: '', log file: '/bazel-scratch/.cache/bazel/_bazel_prow/48d5366022b4e3197674c8d6e2bee219/server/jvm.out')
```

Looks like in the past this has indicated OOM killer (e.g., https://github.com/bazelbuild/bazel/issues/3020). Are we seeing this with other jobs - CI machines running low on memory?

johnSchnake commented:
I've been quite frustrated with the flakes as well; I wasn't sure if the gce issue was a known issue impacting others.

I'd be a somewhat (but not entirely) surprised that OOM is a problem; I'm able to run this bazel job on my laptop while multitasking just fine. And it seems like its failing at another part entirely that wasn't related to my changes. Not sure if/how it is rippling to impact this part of the build.

johnbelamaric commented:
@BenTheElder e2e-kind failed two tests, I believe both due to timeouts of image pulls.

kubelet test: see [here](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/85000/pull-kubernetes-e2e-kind/1225499204302934017#1:build-log.txt%3A5149) and [a few lines later](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/85000/pull-kubernetes-e2e-kind/1225499204302934017#1:build-log.txt%3A5172)

PQDN test: [here](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/85000/pull-kubernetes-e2e-kind/1225499204302934017#1:build-log.txt%3A13655)

Couple things:
 - why is timeout only 60 seconds when there is an image pull involved?
 - why is the DNS test so complicated, it pulls multiple images, writes files, then checks file contents. why not just execute the queries and check them directly? (that's a question for sig-net, I will follow up on it)

The GCE test didn't even get that far, maybe someone with more context can explain why this [timed out](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/85000/pull-kubernetes-e2e-gce/1225499204294545409/#1:build-log.txt%3A326) after more than an hour?

BenTheElder commented:
we don't configure any image pull timeout in KIND, 60s is probably something in kubelet or containerd.

I'm not sure why the DNS test is so complex, in general tests are consolidating on a single e2e test image (`agnhost`) partly to ease porting tests between linux & windows.

BenTheElder commented:
https://github.com/kubernetes/kubernetes/issues/87441 for GCE timeout (staging)

liggitt commented:
test failure looks legitimate
/lgtm cancel
to prevent automated retry

Jefftree commented:
This has been merged with https://github.com/kubernetes/kubernetes/pull/88486

/close
